### ISSUE TITLE
No warning for use of an obsolete constructor as a constructor-constrained type parameter

### ISSUE THREAD

**Issue by alrz** [dotnet/roslyn#19430]
**Version Used**: current

**Steps to Reproduce**:
```cs
public class C<T> where T : new() {
    [Obsolete]
    public C() => new C<C<T>>();
}
```
**Expected Behavior**: warning: C() is obsolete

**Actual Behavior**: no warnings

Related to warning waves

**Comment by jcouv** [comment]
Relates to warning waves https://github.com/dotnet/roslyn/issues/1580
@cston Can you comment on whether this would be desirable behavior for Obsolete?

**Comment by sharwell** [comment]
@alrz It appears you carefully crafted your example to really minimize the amount of code required to demonstrate the problem. However, there's a slight problem which turns out to have a big impact and highlights an edge case - the use of `C<C<T>>` appears inside a method which is marked obsolete. Warnings for obsolete methods are not reported when the use is itself enclosed in a construct which is marked obsolete.

In the original example, I would not expect a warning to appear. However, if the type `C<C<T>>` was used in code which was not obsolete, I might expect a warning.

This is somewhat related to dotnet/roslyn-analyzers#1158.

**Comment by alrz** [comment]
@sharwell yeah, thanks! I'll update the example; just couldn't help it to not write it that way 😄

**Comment by sharwell** [comment]
@alrz Can you bring back the original example as well since it will be important to include it as a test case if we implement this feature?

**Comment by alrz** [comment]
this also happens when an obsolete method implicitly implements an interface member. The warning is lost after type substitution (could be detected at the call-site)

```cs
interface I { void M(); }
class X : I
{
  [Obsolete]
  public void M() {}

  static void Test<T>(T x) where T : I => Test(new X()); // warn?
}

```

**Comment by RikkiGibson** [comment]
It feels like some patterns around testing MEF components depend on the lack of warning for this. e.g. code fix verifiers which often take code fixers with "obsolete" constructors as type arguments:

https://github.com/dotnet/roslyn-sdk/blob/42937ae4bb059b84adcd6109b17c2c70689021db/src/Microsoft.CodeAnalysis.Testing/Microsoft.CodeAnalysis.CSharp.CodeFix.Testing/CSharpCodeFixVerifier%603.cs#L13

**Comment by sharwell** [comment]
@RikkiGibson Yes, that's correct. I've known about this scenario for a while; it was my hope to replace the use of `Obsolete` on these parts with a new set of analyzers:

https://github.com/dotnet/roslyn-analyzers/issues/5358

During the development of these new analyzers, we could find a solution for `CodeFixProvider`. Another option would be modifying the current rules for marking constructors obsolete to 1) not require `CodeFixProvider` to be marked `Obsolete` and 2) report a new warning if the constructor for `CodeFixProvider` has any arguments (the path which creates these objects does not support arguments).
